### PR TITLE
Add image comparer test helper

### PR DESCRIPTION
**Pull request by tocsoft**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/JimBobSquarePants/ImageSharp/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practise as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description
This adds a helper image comparer class to testing to help us turn some of our none tests into real unit tests.

Add a couple of png smoke tests that encode and then decode an image in png then does a visual compare to verify they render the same as an example on how to use the comparer.

### CODE REVIEW

**Comment by antonfirsov** [comment]
Finally, something will prevent me from adding bugs! ^^

**Comment by tocsoft** [comment]
I thought you might appreciate this PR 😉 

I can't promise its the best comparer but its a start and we can always tweek the internals of it now it exists.

File: tests/ImageSharp.Tests/ImageComparer.cs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    using System;
+    using ImageSharp;
+    using Xunit;
+
+    /// <summary>
+    /// Class to perform simple image comparisons.
+    /// </summary>
+    public static class ImageComparer
+    {
+        const int DefaultScalingFactor = 32;
```
Comment (antonfirsov) [inline tests/ImageSharp.Tests/ImageComparer.cs]: Can you add some explanation, links, etc. about how these factors work?

File: tests/ImageSharp.Tests/Formats/Png/PngSmokeTests.cs [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        {
+            // does saving a file then repoening mean both files are identical???
+            using (Image<TColor> image = provider.GetImage())
+            using (MemoryStream ms = new MemoryStream())
+            {
+                // image.Save(provider.Utility.GetTestOutputFileName("bmp"));
+
+                image.Save(ms, new PngEncoder());
+                ms.Position = 0;
+                using (Image img2 = new Image(ms, new Configuration(new PngFormat())))
```
Comment (antonfirsov) [inline tests/ImageSharp.Tests/Formats/Png/PngSmokeTests.cs]: Seems, you will be confilcting with yourself here :(

Comment (tocsoft) [inline tests/ImageSharp.Tests/Formats/Png/PngSmokeTests.cs]: what do you mean?

Comment (tocsoft) [inline tests/ImageSharp.Tests/Formats/Png/PngSmokeTests.cs]: oh right the other PR yeah i know...which ever one in merged first i'll fix the other one before the merge

File: tests/ImageSharp.Tests/ImageComparer.cs (hunk 1 above)
Comment (tocsoft) [inline tests/ImageSharp.Tests/ImageComparer.cs]: I've added some comments to the methods.

File: src/ImageSharp/Formats/Png/PngInterlaceMode.cs [hunk 3]
```diff
@@ -13,11 +13,11 @@ public enum PngInterlaceMode : byte
         /// <summary>
         /// Non interlaced
         /// </summary>
-        None,
+        None = 0,
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Png/PngInterlaceMode.cs]: Good thinking

File: tests/ImageSharp.Tests/Formats/Png/PngEncoderTests.cs [hunk 4]
```diff
@@ -50,5 +51,31 @@ public void ImageCanSavePngInParallel()
                         }
                     });
         }
+
+        [Theory]
+        [WithBlankImages(1, 1, PixelTypes.All)]
+        public void WritesFileMarker<TColor>(TestImageProvider<TColor> provider)
```
Comment (JimBobSquarePants) [inline tests/ImageSharp.Tests/Formats/Png/PngEncoderTests.cs]: Format tests Hurrah! 🍾

File: tests/ImageSharp.Tests/Formats/Png/PngSmokeTests.cs (hunk 2 above)
Comment (JimBobSquarePants) [inline tests/ImageSharp.Tests/Formats/Png/PngSmokeTests.cs]: We'll try and get the other merged first. Just added a  review now.

File: tests/ImageSharp.Tests/TestUtilities/ImageProviders/TestPatternProvider.cs [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+                    topBand = topBand + height;
+                    c.PackFromVector4(blue);
+                    for (int y = topBand; y < bottom; y++)
+                    {
+                        pixels[x, y] = c;
+                    }
+                }
+
+            }
+            private static void Raninbow(PixelAccessor<TColor> pixels)
```
Comment (JimBobSquarePants) [inline tests/ImageSharp.Tests/TestUtilities/ImageProviders/TestPatternProvider.cs]: Rainbow

File: tests/ImageSharp.Tests/TestUtilities/ImageProviders/TestPatternProvider.cs [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+// Licensed under the Apache License, Version 2.0.
+// </copyright>
+
+namespace ImageSharp.Tests
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Numerics;
+
+    public abstract partial class TestImageProvider<TColor>
```
Comment (JimBobSquarePants) [inline tests/ImageSharp.Tests/TestUtilities/ImageProviders/TestPatternProvider.cs]: Can I get some descriptions on the private methods here. Just to ensure everyone can easily see exactly what we are testing.

File: tests/ImageSharp.Tests/Formats/Png/PngSmokeTests.cs [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    using ImageSharp.IO;
+
+    public class PngSmokeTests
+    {
+        [Theory]
+        [WithTestPatternImages(300, 300, PixelTypes.All)]
+        public void GeneralTest<TColor>(TestImageProvider<TColor> provider)
+            where TColor : struct, IPixel<TColor>
+        {
+            // does saving a file then repoening mean both files are identical???
```
Comment (JimBobSquarePants) [inline tests/ImageSharp.Tests/Formats/Png/PngSmokeTests.cs]: Not sure I follow here but pixel data will be different in jpeg since it's a lossy format.

Comment (antonfirsov) [inline tests/ImageSharp.Tests/Formats/Png/PngSmokeTests.cs]: In the case of Jpeg-s we can apply a higher percentage threshold with `ImageComparer.VisualComparer()`.

File: tests/ImageSharp.Tests/ImageComparer.cs [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+        /// </param>
+        /// <param name="segmentThreshold">
+        /// The threashold of the individual segments before it acumulates towards the overall difference.
+        /// The default undefined value is <see cref="ImageComparer.DefaultSegmentThreshold"/>
+        /// </param>
+        /// <param name="scalingFactor">
+        /// This is a sampling factor we sample a grid of average pixels <paramref name="scalingFactor"/> width by <paramref name="scalingFactor"/> high
+        /// The default undefined value is <see cref="ImageComparer.DefaultScalingFactor"/>
+        /// </param>
+        public static void VisualComparer<TColorA, TColorB>(Image<TColorA> expected, Image<TColorB> actual, float imageTheshold = DefaultImageThreshold, byte segmentThreshold = DefaultSegmentThreshold, int scalingFactor = DefaultScalingFactor)
```
Comment (antonfirsov) [inline tests/ImageSharp.Tests/ImageComparer.cs]: Shouldn't this be named differently? eg. `CheckSimilarity()`
It does an assertion!

File: tests/ImageSharp.Tests/Formats/Png/PngSmokeTests.cs (hunk 7 above)
Comment (antonfirsov) [inline tests/ImageSharp.Tests/Formats/Png/PngSmokeTests.cs]: In the case of Jpeg-s we can pass a higher threshold to `ImageComparer.VisualComparer()`.

Comment (JimBobSquarePants) [inline tests/ImageSharp.Tests/Formats/Png/PngSmokeTests.cs]: True dat.

File: tests/ImageSharp.Tests/ImageComparer.cs (hunk 8 above)
Comment (JimBobSquarePants) [inline tests/ImageSharp.Tests/ImageComparer.cs]: Yeah, that'd be much more explicit.

File: tests/ImageSharp.Tests/TestUtilities/ImageProviders/BlankProvider.cs [hunk 9]
```diff
@@ -6,25 +6,46 @@
...
+            public override void Deserialize(IXunitSerializationInfo info)
+            {
+                this.Width = info.GetValue<int>("width");
+                this.Height = info.GetValue<int>("height");
+                base.Deserialize(info);
+            }
+
+            public override void Serialize(IXunitSerializationInfo info)
+            {
+                info.AddValue("width", this.Width);
```
Comment (antonfirsov) [inline tests/ImageSharp.Tests/TestUtilities/ImageProviders/BlankProvider.cs]: Sometimes theory arguments work without being explicitly `IXunitSerializable`, sometimes they don't. 
Have you figured out the magic parts? :)

Comment (tocsoft) [inline tests/ImageSharp.Tests/TestUtilities/ImageProviders/BlankProvider.cs]: not really just started adding stuff till it started working correctly.

**Comment by tocsoft** [comment]
everyone happy with this now?

**Comment by JimBobSquarePants** [comment]
:shipit:

**Pull request merged**